app/key_loader.py

```diff
@@ -0,0 +1,71 @@
+from abc import ABC, abstractmethod
+import ssl
+import os
+import re
+from tempfile import NamedTemporaryFile
+
+from app.models import TLS
+
+
+OB_CERTS_DIR = os.path.abspath(
+    os.environ.get("OB_CERTS_DIR", "/app/open_banking_certs")
+)
+PASSWORD_SUFFIX = "_PASSWORD"
+CREDENTIAL_SUFFIX = "_CRED"  # used for both public certificates and private keys
+
+
+def read_key_password(key_path: str) -> str | None:
+    return os.environ.get(re.sub(r"[\-\.\/]", "_", key_path).upper() + PASSWORD_SUFFIX)
+
+
+class KeyLoader(ABC):
+    @abstractmethod
+    def get_content(self, name: str) -> bytes: ...
+
+    @abstractmethod
+    def update_ssl_context(
+        self, context: ssl.SSLContext, tls: TLS
+    ) -> ssl.SSLContext: ...
+
+
+class FileKeyLoader(KeyLoader):
+    def _get_ob_certs_file_path(self, path: str) -> str:
+        abspath = os.path.abspath(os.path.join(OB_CERTS_DIR, path))
+        if os.path.commonpath([OB_CERTS_DIR, abspath]) != OB_CERTS_DIR:
+            raise ValueError(
+                f"{path} is not inside open banking certificates directory"
+            )
+        return abspath
+
+    def get_content(self, name: str) -> bytes:
+        with open(self._get_ob_certs_file_path(name), "rb") as f:
+            return f.read()
+
+    def update_ssl_context(self, context: ssl.SSLContext, tls: TLS) -> ssl.SSLContext:
+        context.load_cert_chain(
+            self._get_ob_certs_file_path(tls.cert_path),
+            self._get_ob_certs_file_path(tls.key_path),
+            read_key_password(tls.key_path),
+        )
+
+        return context
+
+
+class EnvKeyLoader(KeyLoader):
+    def get_content(self, name: str) -> bytes:
+        if not name.endswith(CREDENTIAL_SUFFIX):
+            name += CREDENTIAL_SUFFIX
+        return os.environ[name].encode()
+
+    def update_ssl_context(self, context: ssl.SSLContext, tls: TLS) -> ssl.SSLContext:
+        with NamedTemporaryFile() as cert_file, NamedTemporaryFile() as key_file:
```

Comment:
would it make sense to do this once at the load of the service? or is it too hard to achieve? writing to file system on every request feels like a heavy operation (especially that we need to flush), but I'm not sure, OS might be optimised for this.

Reply:
We could do that, but I thought that this may raise some security concerns. Perhaps someone wants to keep the file in-memory only.

On my laptop writing a file 1000 times takes ~0.3s. => 0.3ms per file.    
`.flush()` call does not seem to affect the speed.

Reply:
Considering that the sensitive key files are encrypted (if they are not, there should be security concerns of a greater magnitude), it should be safe to put them on the filesystem.

But yeah, you seem right with your measures and 0.3 (or even 1, or even 10) ms probably don't matter much, considering that request latency is anyway over 100 ms. Only if this operation will somehow block threads, which it probably not the case.